### PR TITLE
vmm: Detect and handle AMD SME (Secure Memory Encryption)

### PR DESCRIPTION
**Pull request by crab2313**
Some physical address bits may become reserved in page table when SME is enabled on AMD platform. Guest will trigger a reserved bit violation page fault in this case due to write these reserved bits to 1 in page table. We need reduce the reserved bits to get the right physical address range.

### CODE REVIEW

File: vmm/src/vm.rs [hunk 1]
```diff
@@ -213,9 +213,20 @@ fn get_host_cpu_phys_bits() -> u8 {
     unsafe {
         let leaf = x86_64::__cpuid(0x8000_0000);
 
+        let reduced = if leaf.eax >= 0x8000001f
```
Comment (rbradford) [inline vmm/src/vm.rs]: I think this needs some comments (perhaps referencing a spec.)

File: vmm/src/vm.rs [hunk 2]
```diff
@@ -213,9 +213,20 @@ fn get_host_cpu_phys_bits() -> u8 {
...
+            && leaf.ebx == 0x68747541
```
Comment (rbradford) [inline vmm/src/vm.rs]: From https://www.kernel.org/doc/Documentation/x86/amd-memory-encryption.txt
```
Support for SME and SEV can be determined through the CPUID instruction. The
CPUID function 0x8000001f reports information related to SME:

	0x8000001f[eax]:
		Bit[0] indicates support for SME
		Bit[1] indicates support for SEV
	0x8000001f[ebx]:
		Bits[5:0]  pagetable bit number used to activate memory
			   encryption
		Bits[11:6] reduction in physical address space, in bits, when
			   memory encryption is enabled (this only affects
			   system physical addresses, not guest physical
			   addresses)
```

I don't see any mention of ecx or edx.

File: vmm/src/vm.rs (hunk 1 above)
Comment (crab2313) [reply vmm/src/vm.rs]: ok. I will add them.

File: vmm/src/vm.rs (hunk 2 above)
Comment (crab2313) [reply vmm/src/vm.rs]: What I do here is checking the vendor string of amd.

0x68747541 Auth  0x444d4163 cAMD  0x69746e65 enti

**Pull request merged**